docs/source/best-practices/checkpointing.rst

```diff
@@ -0,0 +1,162 @@
+
+To submit these jobs to the scheduler, use the command::
+
+   sbatch submit_mdrun_array.bash
+
+Python TensorFlow Checkpointing example
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This example demonstrates how implement a longer TensorFlow ML training using the **tf.keras** Checkpointing `API <https://www.tensorflow.org/tutorials/keras/save_and_load>`_ and multiple shorter Slurm job arrays on the gpu partition.
+Below the example **submit_tf_array.bash** script::
+
+ #!/bin/bash
+ #SBATCH --job-name=myrun
+ #SBATCH --time=00:10:00
+ #SBATCH --partition=gpu
+ #SBATCH --nodes=1
+ #SBATCH --gres=gpu:1
+ #SBATCH --mem=10Gb
+ #SBATCH --output=%A-%a.out
+ #SBATCH --error=%A-%a.err
+ #SBATCH --array=1-10%1  #execute 10 array jobs, 1 at a time.
+
+ module load miniconda3/2020-09
+ source activate tf_gpu
+
+ ##Define the number of steps based on the job id:
+ numOfSteps=$(( 500 * SLURM_ARRAY_TASK_ID ))
+
+ # run the python code, save all output to a log file corresponding the the current job task that is running:
+ python train_with_checkpoints.py $numOfSteps &> log.$SLURM_ARRAY_TASK_ID
+
+Where the checkpointing implementation is given in this code snippet of `train_with_checkpoints.py`::
+
+ checkpoint_path = "training_2/{epoch:d}.ckpt"
+ checkpoint_dir = os.path.dirname(checkpoint_path)
+ cp_callback = tf.keras.callbacks.ModelCheckpoint(
+    filepath=checkpoint_path,
+    verbose=1,
+    save_weights_only=True,
+    period=5)
+
+The full scripts be found `here <https://github.com/NURC-Training/checkpointing-102021/tree/main/Exercise_2>`_ and were modified from `TensorFlow Save and load models <https://www.tensorflow.org/tutorials/keras/save_and_load>`_.
+
+The Slurm option `--array=1-10%1` will create 10 Slurm array tasks, and will run one task job at a time. Note that the saved variable `%A` denotes the main job ID, while variable `%a` denotes the task ID (spanning values 1-10).
+The Shell variable SLURM_ARRAY_TASK_ID denotes the unique task job value. Note that also the output/error files are unique in order to prevent different jobs writing to the same files.
+
+To submit these jobs to the scheduler, use the command::
+   
+  sbatch submit_tf_array.bash
+
+Checkpointing using DMTCP
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+`DMTCP <https://dmtcp.sourceforge.io/>`_ (Distributed MultiThreaded Checkpointing) is a Checkpointing tool that lets you Checkpoint without the need to change your code. It Works with most Linux applications such as Python, Matlab, R, GUI, MPI etc. 
+The program runs in the background of your program, without significant performance loss, and saves the process states into checkpoint files. DMTCP is available on the cluster ::
```

Comment:
you can finish this sentence by saying "and can be accessed as follows:" or something similar